cds/container/hopscotch_hashmap.h

```diff
@@ -0,0 +1,660 @@
+			Otherwise, the functor \p func is called with item found.
+			The functor \p func signature is:
+			\code
+			struct my_functor {
+			void operator()( bool bNew, value_type& item );
+			};
+			\endcode
+			with arguments:
+			- \p bNew - \p true if the item has been inserted, \p false otherwise
+			- \p item - an item of the map for \p key
+
+			Returns std::pair<bool, bool> where \p first is \p true if operation is successful,
+			i.e. the node has been inserted or updated,
+			\p second is \p true if new item has been added or \p false if the item with \p key
+			already exists.
+			*/
+			template <typename K, typename V, typename Func>
+			std::pair<bool, bool> update(K const& key, V const& val, Func func, bool bAllowInsert = true)
+			{
+				bool inserted = false;
+
+				if (!contains(key)) {
+					if (bAllowInsert) {
+						if (insert(key, val)) {
+							inserted = true;
+						}
+						else {
+							return make_pair(false, false);
+						}
+					}
+					else {
+						return make_pair(false, false);
+					}
+				}
+
+				unsigned int hash = calc_hash(key);
+				Bucket* start_bucket = segments_arys + hash;
+				unsigned int try_counter = 0;
+				unsigned int timestamp;
+				do {
+					timestamp = start_bucket->_timestamp;
+					unsigned int hop_info = start_bucket->_hop_info;
+					Bucket* check_bucket = start_bucket;
+					unsigned int temp;
+					for (int i = 0; i < HOP_RANGE; i++) {
+						temp = hop_info;
+						temp = temp >> i;
+
+						if (temp & 1) {
+							if (key == *(check_bucket->_key)) {
+								*(check_bucket->_data) = val;
+								func(*(check_bucket->_data));
+								return make_pair(true, inserted);
+							}
+						}
+						++check_bucket;
+					}
+					++try_counter;
+				} while (timestamp != start_bucket->_timestamp && try_counter < MAX_TRIES);
+
+				if (timestamp != start_bucket->_timestamp) {
+					Bucket* check_bucket = start_bucket;
+					for (int i = 0; i < HOP_RANGE; i++) {
+						if (key == *(check_bucket->_key)) {
+							*(check_bucket->_data) = val;
+							func(*(check_bucket->_data));
+							return make_pair(true, inserted);
+						}
+						++check_bucket;
+					}
+				}
+				return make_pair(false, inserted);
+			}
+
+			template<typename K, typename V>
+			V get(K const& key) {
+				return get(key, [=](K const& one, K const& two) { return one != two; });
+			}
+
+			template<typename K, typename V, typename Pred>
+			V get(K const& key, Pred cmp) {
+				unsigned int hash = calc_hash(key);
+				Bucket* start_bucket = segments_arys + hash;
+				unsigned int try_counter = 0;
+				unsigned int timestamp;
+				do {
+					timestamp = start_bucket->_timestamp;
+					unsigned int hop_info = start_bucket->_hop_info;
+					Bucket* check_bucket = start_bucket;
+					unsigned int temp;
+					for (int i = 0; i < HOP_RANGE; i++) {
+						temp = hop_info;
+						temp = temp >> i;
+
+						if (temp & 1) {
+							if (cmp(key, *(check_bucket->_key)) == 0) {
+								return *(check_bucket->_data);
```

Comment:
This lines are incorrect, again a bad holdover from the initial version. It could be the case that in between reading a matching key and then reading the data at that address that someone could have deleted that bucket and inserted a new item there - therefore reading the incorrect data from the table. This is because reading both the key and the value is not an atomic operation and can change during each.

Reply:
@DaKellyFella, see it. If we stay with hop_info-based implementation, may be we can do optimistic synchronization in current bucket?

Reply:
@LeoSko 

Yup! You can read at will and then check the timestamp at the very end. It relates to the comment above. Once the before and after timestamps match you know for sure everything matches up.